osu.Game/Overlays/NotificationManager.cs

```diff
@@ -0,0 +1,111 @@
+
+        private ScrollContainer scrollContainer;
+        private FlowContainer<NotificationSection> sections;
+
+        [BackgroundDependencyLoader(permitNulls: true)]
+        private void load(OsuColour colours)
+        {
+            Width = width;
+            RelativeSizeAxes = Axes.Y;
+
+            Children = new Drawable[]
+            {
+                new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = Color4.Black,
+                    Alpha = 0.6f,
+                },
+                scrollContainer = new ScrollContainer()
+                {
+                    Margin = new MarginPadding { Top = Toolbar.Toolbar.HEIGHT },
+                    Children = new[]
+                    {
+                        sections = new FlowContainer<NotificationSection>
+                        {
+                            Direction = FlowDirection.VerticalOnly,
+                            AutoSizeAxes = Axes.Y,
+                            RelativeSizeAxes = Axes.X,
+                            Children = new []
+                            {
+                                new NotificationSection
+                                {
+                                    Title = @"Notifications",
+                                    ClearText = @"Clear All",
+                                    AcceptTypes = new [] { typeof(SimpleNotification) },
+                                },
+                                new NotificationSection
+                                {
+                                    Title = @"Running Tasks",
+                                    ClearText = @"Cancel All",
+                                    AcceptTypes = new [] { typeof(ProgressNotification) },
+                                },
+                            }
+                        }
+                    }
+                }
+            };
+        }
+
+        int runningDepth = 0;
+
+        public void Post(Notification notification)
+        {
+            ++runningDepth;
+            notification.Depth = notification.DisplayOnTop ? runningDepth : -runningDepth;
+
+            var hasCompletionTarget = notification as IHasCompletionTarget;
+            if (hasCompletionTarget != null)
+                hasCompletionTarget.CompletionTarget = Post;
+
+            var ourType = notification.GetType();
+            sections.Children.FirstOrDefault(s => s.AcceptTypes.Any(accept => ourType == accept || ourType.IsSubclassOf(accept)))?.Add(notification);
+        }
+
+        protected override void PopIn()
+        {
+            base.PopIn();
+
+            scrollContainer.MoveToX(0, TRANSITION_LENGTH, EasingTypes.OutQuint);
+            MoveToX(0, TRANSITION_LENGTH, EasingTypes.OutQuint);
+            FadeTo(1, TRANSITION_LENGTH / 2);
+        }
+
+        private void markAllRead()
+        {
+            sections.Children.ForEach(s => s.MarkAllRead());

```

Comment:
case-by-case. foreach in performance places to avoid delegate allocs (this isn't such a case).